Port to Solaris

Currently mio is using for Solaris epoll interface. But Solaris doesn't have this interface. Only Illumos has it.

There are two things to be done:
- To distinguish between Solaris and Illumos.
- Implement selector on Solaris (probably via even ports API)

> Can you try testing Mio (by running `cargo test`) and give us an output of the error, then we can try and fix it.

```
error: linking with `gcc` failed: exit code: 1
  |
  = note: "gcc" "-m64" "-L" "/builds/psumbera/rustc-1.37.0/lib/rustlib/sparcv9-sun-solaris/lib" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.11v52vx52e6g1ahq.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.13gmn7nbr8fz58ef.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.1a0zdiqerop42dtc.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.1qz1pata0l5g4pvq.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.1vxfmru39gxz4ar3.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.1xciivo4r5pr1dsq.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.24zcpn44b12eo1td.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.25r1lhkoly7a984g.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.2l3vzd2qkdu4ithc.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.2rlwld8q5u3rzw4m.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.2zklttukxouc1zzb.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.348stzlifohl7ig9.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.375xutsz6sjyrar4.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.39xu1wql7p9t2w6.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.4ap6wzhjmmr89gni.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.4dgjk7iptqfgxwiu.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.4kqcii0nyyrp2t8o.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.4txpvbe1msmro5ki.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.4zymg0dpk7z686tz.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.503g90dslmwg1yit.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.58y8hj1dfye6cfhp.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.f7q89bdlkfageun.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.jtqk5a7j6hdi865.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.no53nngwdccutsq.rcgu.o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.wwpf7khpch25zyy.rcgu.o" "-o" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc" "/builds/psumbera/mio-test/target/debug/examples/udp_server-2bc75a83567111bc.2c8h1obrgnxnb2gw.rcgu.o" "-Wl,-zignore" "-nodefaultlibs" "-L" "/builds/psumbera/mio-test/target/debug/deps" "-L" "/builds/psumbera/rustc-1.37.0/lib/rustlib/sparcv9-sun-solaris/lib" "-Wl,-Bstatic" "/builds/psumbera/mio-test/target/debug/deps/libmio-13bc6fc3f9b5e16b.rlib" "/builds/psumbera/mio-test/target/debug/deps/liblibc-96233ac8f4948085.rlib" "/builds/psumbera/mio-test/target/debug/deps/liblog-56b3bee101ff990c.rlib" "/builds/psumbera/mio-test/target/debug/deps/libcfg_if-f7c1f14feae8d52e.rlib" "-Wl,--start-group" "/builds/psumbera/rustc-1.37.0/lib/rustlib/sparcv9-sun-solaris/lib/libstd-8db4654fdc908441.rlib" "/builds/psumbera/rustc-1.37.0/lib/rustlib/sparcv9-sun-solaris/lib/libpanic_unwind-3bfcf78ba2f0a654.rlib" "/builds/psumbera/rustc-1.37.0/lib/rustlib/sparcv9-sun-solaris/lib/libbacktrace-7e6fbe207ebada1c.rlib" "/builds/psumbera/rustc-1.37.0/lib/rustlib/sparcv9-sun-solaris/lib/libbacktrace_sys-37dec1614f68e8aa.rlib" "/builds/psumbera/rustc-1.37.0/lib/rustlib/sparcv9-sun-solaris/lib/librustc_demangle-7a6232bee123d2d4.rlib" "/builds/psumbera/rustc-1.37.0/lib/rustlib/sparcv9-sun-solaris/lib/libhashbrown-4b510aad541d18fe.rlib" "/builds/psumbera/rustc-1.37.0/lib/rustlib/sparcv9-sun-solaris/lib/librustc_std_workspace_alloc-58508acde8e63d53.rlib" "/builds/psumbera/rustc-1.37.0/lib/rustlib/sparcv9-sun-solaris/lib/libunwind-766e9e902a9094b7.rlib" "/builds/psumbera/rustc-1.37.0/lib/rustlib/sparcv9-sun-solaris/lib/libcfg_if-2e9ddb3347410ede.rlib" "/builds/psumbera/rustc-1.37.0/lib/rustlib/sparcv9-sun-solaris/lib/liblibc-ca073c74dcf43c77.rlib" "/builds/psumbera/rustc-1.37.0/lib/rustlib/sparcv9-sun-solaris/lib/liballoc-92a9c38060cc44cf.rlib" "/builds/psumbera/rustc-1.37.0/lib/rustlib/sparcv9-sun-solaris/lib/librustc_std_workspace_core-953de5af53a4fd9f.rlib" "/builds/psumbera/rustc-1.37.0/lib/rustlib/sparcv9-sun-solaris/lib/libcore-5fd40412d92f6541.rlib" "-Wl,--end-group" "/builds/psumbera/rustc-1.37.0/lib/rustlib/sparcv9-sun-solaris/lib/libcompiler_builtins-ba711d886e9bdfdf.rlib" "-Wl,-Bdynamic" "-lsocket" "-lposix4" "-lpthread" "-lresolv" "-lgcc_s" "-lc" "-lm" "-lrt" "-lpthread"
  = note: Undefined            first referenced
           symbol                  in file
          epoll_wait                          /builds/psumbera/mio-test/target/debug/deps/libmio-13bc6fc3f9b5e16b.rlib(mio-13bc6fc3f9b5e16b.5d4rzgaw272n8cti.rcgu.o)
          epoll_ctl                           /builds/psumbera/mio-test/target/debug/deps/libmio-13bc6fc3f9b5e16b.rlib(mio-13bc6fc3f9b5e16b.5d4rzgaw272n8cti.rcgu.o)
          epoll_create1                       /builds/psumbera/mio-test/target/debug/deps/libmio-13bc6fc3f9b5e16b.rlib(mio-13bc6fc3f9b5e16b.5d4rzgaw272n8cti.rcgu.o)
          ld: fatal: symbol referencing errors
          collect2: error: ld returned 1 exit status


error: aborting due to previous error
```

I'm trying to add 'src/sys/unix/selector/evport.rs' which would add support via Solaris even ports API.

My question is also whether it wouldn't be enough to add support for other platforms via select()?

> Also could you help with setting up the CI for Solaris? Going forward we would like to a bit stricter about what platforms we support and having a CI for it is one of them.

This is different problem we are aware of. As for Solaris right now there is for free available only old image of Solaris 11.4. So if you have virtual environment where to run it (e.g. VirtualBox) I can do some setup. But as you can see see I'm now locked with Rust 1.37 till we solve Mio problem...

I have started my work (see here):
https://github.com/psumbera/mio/commit/f6284afe1753b2c23c07d5dd374a969e2f096b82

But now I'm getting:
```
error[E0277]: `*mut std::ffi::c_void` cannot be sent between threads safely
  --> tests/poll.rs:16:5
   |
16 |     assert_send::<Events>();
   |     ^^^^^^^^^^^^^^^^^^^^^ `*mut std::ffi::c_void` cannot be sent between threads safely
   |
   = help: within `libc::unix::solarish::port_event`, the trait `std::marker::Send` is not implemented for `*mut std::ffi::c_void`
   = note: required because it appears within the type `libc::unix::solarish::port_event`
   = note: required because of the requirements on the impl of `std::marker::Send` for `std::ptr::Unique<libc::unix::solarish::port_event>`
   = note: required because it appears within the type `alloc::raw_vec::RawVec<libc::unix::solarish::port_event>`
   = note: required because it appears within the type `std::vec::Vec<libc::unix::solarish::port_event>`
   = note: required because it appears within the type `mio::event::events::Events`
note: required by `util::assert_send`
  --> tests/util/mod.rs:110:1
   |
110| pub fn assert_send<T: Send>() {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Any idea about this one? Thank you!

> I guess `port_event` contains a `*mut c_void` field?

You are right:

```
    pub struct port_event {
        pub portev_events: ::c_int,
        pub portev_source: ::c_ushort,
        pub portev_pad: ::c_ushort,
        pub portev_object: ::uintptr_t,
        pub portev_user: *mut ::c_void,
    }
```
```portev_user``` are user data. It doesn't have to be pointer (man page says: "user defined cookie"). It's set via port_associate(). I do set there ```token```. Is that right?

I tried quick hack by modifying it to ```::c_ulong``` in libc rust wrapper and now I get only following error:

```
error: `core::slice::<impl [T]>::len` is not yet stable as a const fn
  --> tests/unix_datagram.rs:21:26
   |
21 | const DATA1_LEN: usize = DATA1.len();
   |                          ^^^^^^^^^^^
   |
   = help: add `#![feature(const_slice_len)]` to the crate attributes to enable
```
Any clue?


> In rust raw pointers don't implement the `Send` trait, which is the trait that defines that a type may be send across threads. You'll need to define your own `Event` type that wraps the `port_event` type and implement `Send` and `Sync` on it:
> 
> ```rust
> pub struct Event(libc::port_event);
> 
> unsafe impl Send for Event {}
> unsafe impl Sync for Event {}
> ```

Sorry I'm no good enough with Rust yet. I tried:

```
--- a/src/sys/unix/selector/evport.rs
+++ b/src/sys/unix/selector/evport.rs
@@ -113,7 +113,11 @@ impl Drop for Selector {
     }
 }

-pub type Event = libc::port_event;
+pub struct Event(libc::port_event);
+
+unsafe impl Send for Event {}
+unsafe impl Sync for Event {}
+
 pub type Events = Vec<Event>;

 pub mod event {
```

And I get now:

```
error[E0308]: mismatched types
  --> src/sys/unix/selector/evport.rs:62:37
   |
62 |         syscall!(port_getn(self.ep, events.as_mut_ptr(), events.capacity() as u32, &mut nget as *mut u32, timeout))
   |                                     ^^^^^^^^^^^^^^^^^^^ expected struct `libc::unix::solarish::port_event`, found struct `sys::unix::selector::evport::Event`
   |
   = note: expected type `*mut libc::unix::solarish::port_event`
              found type `*mut sys::unix::selector::evport::Event`

error[E0609]: no field `portev_user` on type `&sys::unix::selector::evport::Event`
   --> src/sys/unix/selector/evport.rs:128:21
    |
128 |         Token(event.portev_user as usize)
    |                     ^^^^^^^^^^^
```

Can you please hint me little bit more?
 
> Note however you must manually ensure that the `libc::port_event` may be send over thread bounds. If this is not allowed we'll need to discuss if we want to drop the `Send` guarantee from `Event`.

I hope there is no problem. Event port API is marked as thread safe.
Or am I missing something?

> `portev_user` are user data. It doesn't have to be pointer (man page says: "user defined cookie"). It's set via port_associate(). I do set there `token`. Is that right?

Yes then it means you can implement `Send` for it and it can be used in `Token`.

> 
> I tried quick hack by modifying it to `::c_ulong` in libc rust wrapper and now I get only following error:
> 
> ```
> error: `core::slice::<impl [T]>::len` is not yet stable as a const fn
>   --> tests/unix_datagram.rs:21:26
>    |
> 21 | const DATA1_LEN: usize = DATA1.len();
>    |                          ^^^^^^^^^^^
>    |
>    = help: add `#![feature(const_slice_len)]` to the crate attributes to enable
> ```
> 
> Any clue?

My guess it that you're using an older rust compiler version. Can you update to a newer version? Preferably 1.39. If not you can change those constants to literal values (that of the length of `DATA1`) for the time being.

> ```
> error[E0308]: mismatched types
>   --> src/sys/unix/selector/evport.rs:62:37
>    |
> 62 |         syscall!(port_getn(self.ep, events.as_mut_ptr(), events.capacity() as u32, &mut nget as *mut u32, timeout))
>    |                                     ^^^^^^^^^^^^^^^^^^^ expected struct `libc::unix::solarish::port_event`, found struct `sys::unix::selector::evport::Event`
>    |
>    = note: expected type `*mut libc::unix::solarish::port_event`
>               found type `*mut sys::unix::selector::evport::Event`
> 
> error[E0609]: no field `portev_user` on type `&sys::unix::selector::evport::Event`
>    --> src/sys/unix/selector/evport.rs:128:21
>     |
> 128 |         Token(event.portev_user as usize)
>     |                     ^^^^^^^^^^^
> ```
> 
> Can you please hint me little bit more?

The first problem is that now we're using different types. `port_getn` expects a pointer to an array of `port_event` structures, but we pass a pointer to the `Event` structs (which we defined). As these are different types Rust won't allow us to do that. I think the easier way forward is to remove the test in `tests/poll.rs` which caused the problem in the first place. We'll figure out later the best way to provide an `Send` implementation.

So revert the suggestion I made and go back to `Event` being a type alias for `port_event`. This should also resolve the second error.

> > Note however you must manually ensure that the `libc::port_event` may be send over thread bounds. If this is not allowed we'll need to discuss if we want to drop the `Send` guarantee from `Event`.
> 
> I hope there is no problem. Event port API is marked as thread safe.
> Or am I missing something?

This is safe, see the first part of my comment

> All libc functions are exported in the `libc` crate (docs: https://docs.rs/libc/0.2.65/libc). So open would be `libc::open(..)`, read `libc::read(..)` etc.

Sorry, I don't seem to be able even open file...

```
error[E0308]: mismatched types
  --> src/sys/unix/selector/devpoll.rs:24:23
   |
24 |         syscall!(open("/dev/poll", O_RDWR)).map(|ep| Selector {
   |                       ^^^^^^^^^^^ expected i8, found str
   |
   = note: expected type `*const i8`
              found type `&'static str`
```
Also not sure where is open defined. Is it here?
https://github.com/rust-lang/libc/blob/77aa3e536811dcb3a379174f8ff41d49acb36f87/src/unix/mod.rs#L699
(seems to be for macos?)

I'm lost again. Can you please help me wit this?

```
   Compiling mio v0.7.0 (/builds/psumbera/rust-mio)
error[E0599]: no method named `as_ref` found for type `libc::unix::pollfd` in the current scope
  --> src/sys/unix/selector/devpoll.rs:79:14
   |
79 |             .as_ref()
   |              ^^^^^^
```

The relevant code is:
```
    pub fn register(&self, fd: RawFd, token: Token, interests: Interests) -> io::Result<()> {

        let buf = libc::pollfd {
               fd: fd,
               events: interests_to_evport(interests) as i16,
               revents: usize::from(token) as i16,
        };
        let buf = buf
            .as_ref()
            .map(|s| s as *const _)
            .unwrap_or(ptr::null_mut());

        syscall!(write(self.ep, buf, 8)).map(|_| ())
    }
```

Thank you very much for helping me! Now I get this:

```
error[E0308]: mismatched types
  --> src/sys/unix/selector/devpoll.rs:79:33
   |
79 |         syscall!(write(self.ep, &buf, 8)).map(|_| ())
   |                                 ^^^^ expected enum `core::ffi::c_void`, found struct `libc::unix::pollfd`
   |
   = note: expected type `*const core::ffi::c_void`
              found type `&libc::unix::pollfd`
```

--- a/README.md
+++ b/README.md
@@ -152,10 +152,8 @@ This uses the Windows AFD system to access socket readiness events.
 
 ### Unsupported
 
-* Solaris, see [issue #1152]
 * Wine, see [issue #1444]
 
-[issue #1152]: https://github.com/tokio-rs/mio/issues/1152
 [issue #1444]: https://github.com/tokio-rs/mio/issues/1444
 
 ## MSRV Policy

